GB+Tree page reuse

burqen opened the pull request:
**What?**
The GB+Tree need to reuse pages, otherwise it would grow very fast. This PR enables that.

**How?**
Pages that are eligible for reuse, tree-node-page with an older generation or a freelist-node-page that has been used up, will be added to the freelist which is made up of freelist-nodes making up a linked list.

Whenever a new page is needed, the FreelistIdProvider check if there are any pages to reuse and pick one of those if there are, otherwise allocate at end of file.

**Concurrency!**
Reuse of pages comes with a lot of concurrency problems around read, see `SeekCursor`. Most of this PR thus revolves around ensuring SeekCursor behaves correctly. The `GBPTreeIT#shouldReadCorrectlyWhenConcurrentlyInsertingOutOfOrder` test this very aggressively with overlapping insert and read with a super-high checkpoint load.

**Is that all?**
Commit messages of individual commits are very detailed so please have a look at those.

Commit:
Introduce Freelist

All pages that can be reused are added to the freelist, like a linked list
of pages. Even freelistpages themselves are added to the freelist when
they become empty.

The freelist is managed by the FreelistIdProvider

New index state fields:
freeListWritePageId
freeListReadPageId
freeListWritePos
freeListReadPos

tinwelint committed:
Added javadocs and some removal of unnecessary code

Commit:
Use zapPage() to clear reused page in FreeListIdProvider

tinwelint committed:
Removes key comparison optimization in SeekCursor

...which didn't quite work in concurrent scenario with id-reuse.

tinwelint committed:
SeekCursor refactorings

- broke out rereading from node header into method.
- wrote more comments
- simplified some conditions and made some conditions clearer

Commit:
GSPP - Read generation together with pointer

When reading a pointer, sometimes we need to be able to retrieve the generation
of that pointer. We let GSPP store information about how that information can be retrieved
in the most significant 2 bytes of the pointer, as only 6 bytes are used for the
actual page id.

Since GSPP#read now gives more information in the returned long than just
the pointer, callers who receive read results can no longer treat them as
purely pointers.

This is mostly a problem in tests which asserts pointers or otherwise
deals with read results directly.
For the main code this change is pretty small since the code that hands
out these read results with gen information added will also be aware
of read results as arguments, knowing that it's only the low 6B that
represent pointer.

Encoding generation and pointer in a single long avoids garbage.

Use case example:
TreeNode treeNode = ...
long readResult = GenSafePointerPair.read( cursor, stableGen, unstableGen, logicalPos );
if ( GenSafePointerPair.isSuccess( readResult ) )
{
    long pointer = GenSafePointerPair.pointer( readResult );
    long gen = treeNode.pointerGen( cursor, readResult );
}
else
{
    // Failure encoded as before
}

tinwelint committed:
Implements TreeNode#pointerGen

Used to extract generation from a read reult containing a successfully
read pointer.

Depending on if the pointer comes from node header (sibling or newGen)
or if it is a child pointer, TreeNode need to use different methods
for setting the offset before delegating to GSP.

tinwelint committed:
SeekCursor concurrency fix

- Guard for nodes being reused while pinned by SeekCursor
A SeekCursor can read unexpected data caused by node page being
reused in two cases:
1. The page currently pinned is reused -> node will have a changed generation.
2. The page is reused after reading a pointer to it but before "landing" on it
-> node will have a generation larger than that of the followed pointer.

We now guard for this by verifying that generation
of node is not larger than the generation of the pointer that
was followed to reach the node and by verifying that the generation
of the node does not change during the SeekCursor's visit.
IF such a generaton abnormailty is observerd, the seek is restarted
from the root with an updated range starting from the previously returned
key.
Note that it IS valid for node generation to be lower than the followed
pointer generation because the generation of the pointer is updated
when we copy if from an older node, e.g. rightSibling of right node in a split.

ConsistenctChecker is also updated with this generation verification.

- verifyGen is no longer done in TreeNode.goTo for two reasons:
1. Reading generation field here is not valid because it is not always
guarded by a shouldRetry or a write lock.
2. The assumption that pointer generation always match generation
of the node that it points to is false. The correct assumption
is that the node generation can not be larger than the pointer
generation.

- verifyType is no longer done in TreeNode.goTo. See reason nr1 above.

- SeekCursor now reread keyCount on every call to next
Previously weird behaviour was observed where the keyCount was
not updated propery. The reason why is unclear but re-reading
it every time also reduse complexity. Win.

- Fixes issue with rediscovering key in SeekCursor
where there would be a need to rediscover key due to read key being too far
to the left and no result had been returned yet.
Previously this would just set pos to 0. This would be a problem being
too far to the left because this would result in an endless loop where
the cursor would never leave the node and never leave the next() either.

Other changes:
- Add very aggressive read concurrency test
- Extract long running tests from GBPTreeTest to GBPTreeIT
- Add possibility to not print values in TreePrinter
- Fix TreePrinter to extract pointer value using GSPP

Commit:
SeekCursor now restart seek from root if node is not of tree node type

This can happen if a page is reused as a free list node.

Commit:
Fix checkstyle

tinwelint committed:
Implements a TODO about checking node type in CC

Commit:
Add test shouldReadCorrectlyWhenConcurrentlyRemovingOutOfOrder

tinwelint committed:
gbptree ConsistencyChecker checks space usage

checks so that all pages between MIN_TREE_NODE_ID and highest allocated id
are either used by tree nodes in the tree or are added to the free list or
are free list pages.

tinwelint committed:
Extra checks reading nodeType and keyCount from tree nodes

Becuse it doesn't make sense to read anything else after reading nodeType
and seing that it isn't even a tree node.

Also keyCount is affecting other reads, which if read from a free-list node
(which is possible due to id-reuse), may be larger than the size of a node and
then cause that next other read to go out-of-bounds in a PageCursor,
which in the end would cause the out-of-bounds check after do-shouldRetry loop
to throw exception.

Now the read key counts are checked for sanity and treated just like reading
from node of a different type.

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/FreelistNode.java:
```diff
@@ -0,0 +1,98 @@
+        }
+        assertPos( pos );
+        GenSafePointer.assertGenerationOnWrite( unstableGeneration );
+        cursor.setOffset( HEADER_LENGTH + pos * ENTRY_SIZE );
```
would be nice if we will extract `HEADER_LENGTH + pos * ENTRY_SIZE` into separate `calculateOffset` method or similar

tinwelint replied:
👍 going for `entryOffset` to match equivalent method(s) in `TreeNode`

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/FreelistNode.java:
```diff
@@ -0,0 +1,98 @@
+
+class FreelistNode
+{
+    static final int PAGE_ID_SIZE = GenSafePointer.POINTER_SIZE;
```
almost all of those flags can be private

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/FreelistNode.java:
```diff
@@ -0,0 +1,98 @@
+        assertPos( pos );
+        cursor.setOffset( HEADER_LENGTH + pos * ENTRY_SIZE );
+        long generation = getUnsignedInt( cursor );
+        long result = generation <= stableGeneration ? get6BLong( cursor ) : NO_PAGE_ID;
```
result here is redundant

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/FreelistNode.java:
```diff
@@ -0,0 +1,98 @@
+import static org.neo4j.index.gbptree.PageCursorUtil.getUnsignedInt;
+import static org.neo4j.index.gbptree.PageCursorUtil.put6BLong;
+
+class FreelistNode
```
Needs javadocification

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/FreeListIdProvider.java:
```diff
@@ -0,0 +1,351 @@
+
+class FreeListIdProvider implements IdProvider
+{
+    interface Monitor
```
shouldn't this be abstract class instead of interface with all default methods?

tinwelint replied:
I like Monitors as interfaces, this way you have more freedom in how to implement them. Many (most) of monitors in neo4j are interfaces, aren't they?)

MishaDemianenko replied:
for sure lets keep it as interface, np

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/FreeListIdProvider.java:
```diff
@@ -0,0 +1,351 @@
+
+    void initializeAfterCreation() throws IOException
+    {
+        writePageId = readPageId = nextLastId();
```
please use multiple assignments on multiple lines

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/FreeListIdProvider.java:
```diff
@@ -0,0 +1,351 @@
+        {
+            goTo( cursor, "free-list", writePageId );
+            freelistNode.initialize( cursor );
+            checkOutOfBounds( cursor );
```
cursor.shouldRetry loop is missing?

tinwelint replied:
No, because that page is write locked

MishaDemianenko replied:
Well that behaviour is not described in javadoc of page cursor then i would say java doc should be updated to mention this

tinwelint replied:
PageCursor javadoc to be updated separately.

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/FreeListIdProvider.java:
```diff
@@ -0,0 +1,351 @@
+    {
+        try ( PageCursor cursor = pagedFile.io( writePageId, PagedFile.PF_SHARED_WRITE_LOCK ) )
+        {
+            if ( !cursor.next() )
```
cursor.shouldRetry?

tinwelint replied:
No, because that page is write locked

MishaDemianenko replied:
yeah page cursor java doc then need to be updated

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/FreeListIdProvider.java:
```diff
@@ -0,0 +1,351 @@
+            long nextFreelistPage = acquireNewId( stableGeneration, unstableGeneration, false );
+            try ( PageCursor cursor = pagedFile.io( writePageId, PagedFile.PF_SHARED_WRITE_LOCK ) )
+            {
+                if ( !cursor.next() )
```
cursor.shouldRetry?

tinwelint replied:
No, because that page is write locked

MishaDemianenko replied:
yeah page cursor java doc then need to be updated

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/FreeListIdProvider.java:
```diff
@@ -0,0 +1,351 @@
+                {
+                    // FreelistNode compares generation and so this means that we have an available
+                    // id in the free list which we can acquire from a stable generation
+                    if ( ++readPos >= freelistNode.maxEntries() )
```
can we please extract `++` from if statement

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/FreeListIdProvider.java:
```diff
@@ -0,0 +1,351 @@
+        if ( (readPageId != writePageId || readPos < writePos) &&
+                (allowTakeLastFromPage || readPos < freelistNode.maxEntries() - 1) )
+        {
+            // It looks like read pos is < write pos so check if we can grab the next id
```
based on previous if condition this observation is not correct

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/FreeListIdProvider.java:
```diff
@@ -0,0 +1,351 @@
+        visitor.accept( pageId );
+        try ( PageCursor cursor = pagedFile.io( pageId, PagedFile.PF_SHARED_READ_LOCK ) )
+        {
+            PageCursorUtil.goTo( cursor, "free-list", readPageId );
```
since currently pageId equal to readPageId we should be able to move PageCursorUtil.goTo inside a while loop without almost any changes. You will have it as first statement in your while loop instead of last one that will make it a bit more logical and easy to follow

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/FreeListIdProvider.java:
```diff
@@ -0,0 +1,351 @@
+        int pos = readPos;
+        try ( PageCursor cursor = pagedFile.io( pageId, PagedFile.PF_SHARED_READ_LOCK ) )
+        {
+            PageCursorUtil.goTo( cursor, "free-list", readPageId );
```
same thing with goTo and while loop as above

MishaDemianenko commented on community/index/src/test/java/org/neo4j/index/gbptree/FreeListIdProviderTest.java:
```diff
@@ -0,0 +1,286 @@
+        assertTrue( expected.isEmpty() );
+    }
+
+    private void fillPageWithCrapData( long releasedId ) throws IOException
```
i think fillPageWithRandomData is a bit more descriptive name

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/GenSafePointerPair.java:
```diff
@@ -134,77 +170,52 @@ public static long read( PageCursor cursor, long stableGeneration, long unstable
+        long genOffset = isLogicalPos ? logicalPos : gsppOffset;
+        if ( (genOffset & ~MAX_GEN_OFFSET_MASK) != 0 )
+        {
+            throw new IllegalArgumentException( "Illegal genOffset:" + genOffset + ", it would be too large, max is " +
```
here we throw exception inside of cursor.shouldRetry loop and base our checks on possibly non valid data at all, and we should not do that

tinwelint replied:
Not quite true. The exception is never triggered on data that has been read, only data passed in to the method and so is an illegal argument to that method.

MishaDemianenko replied:
`logicalPos` is evaluated based on a data that we read from a cursor and can be incorrect right?

tinwelint replied:
It's not, it's passed in from outside. EDIT: I'm wrong, there's a usage in SeekCursor, let's fix it

tinwelint replied:
hmm, actually, we _are_ safe here because keyCount check above guards for that, `KeySearch.search` cannot go outside 0 and keyCount. SImply extended the keyCount check for negative numbers as well.

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/ConsistencyChecker.java:
```diff
@@ -50,34 +56,147 @@
+     * assert calls.
+     * @throws IOException on {@link PageCursor} error.
+     */
+    public boolean checkSpace( PageCursor cursor, long lastId, PrimitiveLongIterator freelistIds ) throws IOException
```
can we extract some pieces of this huge method into separate methods? like: addRightSiblings, findLeftMostSibling etc

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/ConsistencyChecker.java:
```diff
@@ -105,6 +228,34 @@ else if ( node.isLeaf( cursor ) )
+        long newGen = node.newGen( cursor, stableGeneration, unstableGeneration );
+        if ( TreeNode.isNode( newGen ) )
+        {
+            System.err.println( "WARNING: we ended up on an old generation " + cursor.getCurrentPageId() +
```
why we print only error message here and continue but in all other places with fail with assertion error?

tinwelint replied:
Hmm, perhaps serving as additional debug information. This could be enforced later in the development process instead. Although I can't recall it ever happening.

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/FreeListIdProvider.java:
```diff
@@ -0,0 +1,351 @@
+     */
+    private final Monitor monitor;
+
+    FreeListIdProvider( PagedFile pagedFile, int pageSize, long lastId, Monitor monitor )
```
why do we need to pass lastId as one of the constructor params while it will be reassign when `initialize` will be called?

tinwelint replied:
`initialize` isn't called when creating an index, only consecutive "open"s. `FreeListIdProvider` could be changed to also know which one is the "first" lastId, but currently only GBPTree knows that and passes that knowledge to it, which I don't think is a bad idea.

MishaDemianenko replied:
ok i see i missed the branch with initializeAfterCreation call, its fine then to keep it like it is now

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/GBPTree.java:
```diff
@@ -219,13 +220,33 @@ default void checkpointCompleted()
+     * Odd -> being updated
+     * Even -> stable
+     */
+    private volatile int rootUpdateGuard;
```
this is extremely surprising guard and way of using it. Ideally it should be replaced by one long where root generation and root id combined which updated atomically. Otherwise it should be extracted to separate class that will perform guarding etc and will provide consistent values to the tree.
Visibility of rootId and generation is in question now always and hard to figured out.

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/GBPTree.java:
```diff
@@ -134,6 +135,11 @@ default void checkpointCompleted()
         }
     }
 
+    interface RootCatchup
```
Can this be moved to upper level class, seek cursor using it so it looks like it deserve it's own class.

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/GBPTree.java:
```diff
@@ -594,10 +647,20 @@ boolean consistencyCheck() throws IOException
+            PageCursorUtil.goTo( cursor, "root", rootId );
+
+            PrimitiveLongSet freelistIds = Primitive.longSet();
+            freeList.visitFreelistPageIds( id -> freelistIds.add( id ) );
```
can be method reference

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/gbptree/TreeNode.java:
```diff
@@ -197,6 +218,20 @@ void setNewGen( PageCursor cursor, long newGenId, long stableGeneration, long un
+    {
+        if ( !GenSafePointerPair.isRead( readResult ) )
+        {
+            throw new IllegalArgumentException( "Expected read result, but got " + readResult );
```
It's not always safe to thrown exception from here since it can be based on data that is incorrect since we read it from cursor that should be retried

tinwelint replied:
You're right in reality, but not conceptually, actually. The `isRead` is determined by a bit in the `readResult` which is set by the `GSPP#read` method itself, regardless of what it reads.

With this said I can see that the `GSPP#read` method doesn't control the pointer it puts into the `readResult` and so it may very well be huge (due to invalid read from `PageCursor`) and overflows into that bit too. That should be controlled in that `read` method anyhow. And so I think the action point here is to add that control and then the above check is safe to do.

MishaDemianenko replied:
ok agree

MishaDemianenko reviewed (changes requested):
Some classes are getting quite big code wise and in terms of their responsibilities and it's quite hard to follow what's going on there: for example SeekCursor when you need to scroll to see the end of a loop that's unmanageable complexity.
Also shortage of naming (gen insteand of generation) for example really helping

MishaDemianenko commented:
Most of the errors have something to do with cursor misuse we need to add tests that use Adversarial cursors with high probability of errors to check that we do not have a lot of places where we miss corresponding retry and error checks

MishaDemianenko commented:
Introduced way of page reuse will insert new child node page in any free page in file and that will brake data locality since that place can be potentially be far from siblings, parent or anything.
Not sure that it's something we should change now, but we should be aware about that problem.

tinwelint commented:
@MishaDemianenko much obliged, I think generally you're right in most things here... we'll start addressing them!

tinwelint committed:
Simplifications and documentations in free-list code

tinwelint committed:
Moves management of root id/gen out from GBPTree

out into a Root class. GBPTree now has a volatile reference to a Root which is
reassigned to a new instance when the root changes, which happens rarely.
To read root id + generation a reader can acquire a local reference to the root
field and call id() and generation() on it.

tinwelint commented:
Note on data locality, I don't think data locality changes from w/o free-list since there's no control in which order data is inserted/changed. Also the very nature of GB+Tree is that new generations create new pages and leave the old, so tree "sortedness" doesn't match physical "sortedness" anyway.

tinwelint committed:
Added PageCursorUtilTest for ensuring that 6B put/get doesn't overflow

tinwelint committed:
Checks keyCount sanity for negative numbers as well in SeekCursor

tinwelint commented:
@MishaDemianenko pushed fixes for your comments

MishaDemianenko commented on community/index/src/test/java/org/neo4j/index/gbptree/SeekCursorTest.java:
```diff
@@ -990,6 +995,144 @@ public void shouldFailChildPointerIfReadFailureNotCausedByCheckpoint() throws Ex
+            }
+            catch ( IOException e )
+            {
+                throw new RuntimeException( e );
```
its not that important here, but in general for future cases like that its better to use UncheckedIOException
